### DISCUSSION TITLE
Version listed in version history but not in cache

### DISCUSSION THREAD

**Discussion by ibensalah** [verdaccio/verdaccio#2345] [Q&A] [answered]
Hello, 

I would like to understand how the version history and the cache work. 

Currently, when I go on the package's page on verdaccio I see in version history sections some versions of the package. For exemple I have : 
`v4.9.3`
`v4.9.2`
`v4.9.1`

When I look in the cache, I only have the tarball for the version `4.9.1`. 

The commands bellow are made in an offline environnement 

- If I make `npm install myPackage` it does not work and it says that it can not find the version 4.9.3 of my package.
- If I make an `npm install mypackage@4.9.3` I got an error (which is normal because the package is not in cache)
- If I make an `npm install mypackage@4.9.1 `The package is installed. 

I have read the documentation and my understanding is that if a package is listed in the version history but not available in the cache, verdaccio will try to get it using the uplinks. 

As my environnement is offline, I would like to get the last version available in my cache when I make `npm install myPackage`. 

What do I have to do in my configuration of verdaccio to have my desired behavior. 

Thank you very much !

**Comment by juanpicado** [comment, marked as answer]
I would need to write an article about this, but in an nutshell, even if the version is displayed on the web and via `npm info` does not means the tarball is there. 

There are two fundamental pieces of the cache, the **metadata** (package.Json) and the **tarball**. The metadata is updated constantly and is the full info about that package 📦 all versions included. The tarballs in the same folder  are downloaded by demand, so even if the metadata indicates are 10 versions only 1 tarball has been requested by far.

Hope this helps.

**Reply by ibensalah** [reply]
Thank you for your answer. 

Is there a way to get verdaccio to acts offline ? I don't want npm trying to download a version listed in the metadata but with no tarball in cache. 

Because of this behavior I cannot use verdaccio completely offline. 

Thank you !

**Reply by juanpicado** [reply]
If you want to be offline, remove the `proxy` field on the packages that you don't want to be updated. If you remove all, then is fully offline.

**Reply by ibensalah** [reply]
In order to update the cache I run an `npm install` from a computer that is connected to the internet. Then I copy the updated cache and into the cache that is on the offline computer. (The two computers are different). 

The offline computer has no proxy in its `config.yaml` file but the computer that is connected to the internet has a `proxy:npmjs` field in its` config.yaml` file in order to get the versions from npm when updating the cache.  

 Should I remove  `proxy:npmjs` from the verdaccio instance that is connected to the internet ? I am afraid that I will not be able to update my cache if I do this.

**Reply by juanpicado** [reply]
> Should I remove proxy:npmjs from the verdaccio instance that is connected to the internet ? 
No, only the one is offline.

Seems related to https://github.com/verdaccio/verdaccio/discussions/2306#discussioncomment-1045935
